src/Compilers/CSharp/Test/Syntax/Parsing/ForStatementParsingTest.cs

```diff
@@ -448,4 +443,2690 @@ public void TestVariableDeclaratorVersusCondition1()
+            N(SyntaxKind.ObjectCreationExpression);
+            {
+                N(SyntaxKind.NewKeyword);
+                N(SyntaxKind.IdentifierName);
+                {
+                    N(SyntaxKind.IdentifierToken, "A");
+                }
+                N(SyntaxKind.ObjectInitializerExpression);
+                {
+                    N(SyntaxKind.OpenBraceToken);
+                    N(SyntaxKind.CloseBraceToken);
+                }
+            }
+            N(SyntaxKind.CloseParenToken);
+            N(SyntaxKind.EmptyStatement);
+            {
+                N(SyntaxKind.SemicolonToken);
+            }
+        }
+        EOF();
+    }
+
+    [Fact]
+    public void TestVariousExpressions_Prefix()
+    {
+        UsingStatement("""
+            for (++a;++a;++a);
+            """);
+
+        N(SyntaxKind.ForStatement);
+        {
+            N(SyntaxKind.ForKeyword);
+            N(SyntaxKind.OpenParenToken);
+            N(SyntaxKind.PreIncrementExpression);
+            {
+                N(SyntaxKind.PlusPlusToken);
+                N(SyntaxKind.IdentifierName);
+                {
+                    N(SyntaxKind.IdentifierToken, "a");
+                }
+            }
+            N(SyntaxKind.SemicolonToken);
+            N(SyntaxKind.PreIncrementExpression);
+            {
+                N(SyntaxKind.PlusPlusToken);
+                N(SyntaxKind.IdentifierName);
+                {
+                    N(SyntaxKind.IdentifierToken, "a");
+                }
+            }
+            N(SyntaxKind.SemicolonToken);
+            N(SyntaxKind.PreIncrementExpression);
+            {
+                N(SyntaxKind.PlusPlusToken);
+                N(SyntaxKind.IdentifierName);
+                {
+                    N(SyntaxKind.IdentifierToken, "a");
+                }
+            }
+            N(SyntaxKind.CloseParenToken);
+            N(SyntaxKind.EmptyStatement);
+            {
+                N(SyntaxKind.SemicolonToken);
+            }
+        }
+        EOF();
+    }
+
+    [Fact]
+    public void TestVariousExpressions_Query()
+    {
+        UsingStatement("""
+            for (from a in b select c;from a in b select c;from a in b select c);
+            """,
+            // (1,1): error CS1073: Unexpected token 'from'
+            // for (from a in b select c;from a in b select c;from a in b select c);
+            Diagnostic(ErrorCode.ERR_UnexpectedToken, "for (from a in b select c;").WithArguments("from").WithLocation(1, 1),
+            // (1,1): error CS1003: Syntax error, 'foreach' expected
+            // for (from a in b select c;from a in b select c;from a in b select c);
+            Diagnostic(ErrorCode.ERR_SyntaxError, "for").WithArguments("foreach").WithLocation(1, 1),
+            // (1,18): error CS1026: ) expected
+            // for (from a in b select c;from a in b select c;from a in b select c);
+            Diagnostic(ErrorCode.ERR_CloseParenExpected, "select").WithLocation(1, 18));
```

Comment:
this is surprising, and likely a parsing bug (though practically impossible for it to matter in practice).  It looks like we see `from a` and think that that's the start of a variable declarator (with type `from` and identifier `a`).